README.md
# puppetlabs-servicenow_reporting_integration

The ServiceNow reporting integration module ships with a `servicenow` [report processor](https://puppet.com/docs/puppet/7/reporting_write_processors.html) that can send one of two kinds of information back to Servicenow. It can send events that are handled by Servicenow to create Alerts and Incidents, or you can create Incidents directly.

#### Table of Contents

1. [Pre-reqs](#pre-reqs)
2. [Setup](#setup)
    * [Events](#events)
    * [Incidents](#incidents)
    * [Filtering](#filtering)
3. [Troubleshooting](#troubleshooting)
4. [Development](#development)
5. [Known Issues](#known-issues)

## Pre-reqs

* Puppet Enterprise version 2019.8.1 (LTS) or newer
* A ServiceNow instance (dev or enterprise)

## Setup

1. Install the `puppetlabs-servicenow_reporting_integration` module on your Puppet server.

The module can send events or it can create incidents, but attempting to do both will cause a [catalog compilation](https://puppet.com/docs/puppet/7/subsystem_catalog_compilation.html) failure. Please choose which one of the two you would like to do and only classify your Puppet server nodes with one of those classes, either event management, or incident management.

### Events

To send [events](https://puppet.com/docs/pe/2019.8/analyze_changes_across_runs.html), classify your Puppet servers with the `servicenow_reporting_integration::event_management` class. The minimum parameters you need to configure for the class to work are `instance` (the fqdn of the Servicenow instance), and then `user` and `password`, or you can bypass username/password authentication and use an oauth token using the `oauth_token` parameter.

By default each event will include the following information:
* __Source__: Puppet
* __Node__: The node the agent ran on
* __Type__: The type of event. Can be one of `node_report_intentional_changes`, `node_report_corrective_changes`, `node_report_unchanged`, `node_report_failed`
* __Source instance__: The name of the Puppet server that generated the report
* __Message Key__: A hash of all of the relevant report properties to ensure that future events are grouped together properly
* __Severity__: The highest severity rating of all of the events that occurred in a given run. These severity levels can be configured via the `<change_type>_event_severity` class parameters, including the severity of `no_changes` reports via the `no_changes_events_severity` parameter.
* __Description__: Contains the following:
  * __Report Labels__ - All of the different kinds of events that were in the Puppet run that generated this event. While severity is only the single highest severity, these labels tell you all of the kinds of events that occurred.
  * __Environment__ - The Puppet environment the node is assigned to.
  * __Resource Statuses__ - The full name of the resource, the name of the property and the event message, and the file and line where the resource was defined, for each resource event that was in the report. All resource events are included except for `audit` events, which are resources for which nothing interesting happened; Puppet simply verified that they are currently still in the correct state.
  * __Facts__ - All of the facts that were requested via the `include_facts` parameter. The default format is yaml, but can be changed via the `facts_format` parameter.
* __Additional Information__: The additional information field contains data about the event in JSON format to make it easy to target that information for rules and workflows. It will contain separate top level keys for each of the facts included in the event from the `include_facts` parameter. It will also contain the following keys
  * __environment__: The environment the node is assigned to
  * __report_labels__: A list of all of the different kinds of events that are included in this event.
  * __corrective_changes__: All of the resource events that were triggered by corrective changes.
  * __intentional_changes__: All of the resource events that were triggered by intentional changes.
  * __pending_corrective_changes__: All of the resource events that were trigged by pending (noop) corrective changes.
  * __pending_intentional_changes__: All of the resource events that were triggered by pending (noop) intentional changes.
  * __failures__: All of the resources events that were triggered by resource failures.

By default, the event_creation_conditions is set to ['always']. This means the module will send a single event for every Puppet run on every node. If you would like to filter on the types of reports being sent, the other available types are: `corrective_changes`, `intentional_changes`, `pending_corrective_changes`, `pending_intentional_changes`, and `failures`. You can set the filter in the module's parameters. If nothing interesting such as changes or a failure happened in a given Puppet run then the event type will be `node_report_unchanged`, there will be no resources listed in the description, and the report severity default value will be `OK`.

If a change happens then the event type and severity will be updated, and any resources that changed will be listed in the description.

If multiple events happen e.g. two resources make corrective changes, but a third resource fails, then the report type will be `node_report_failure`, the severity by default will be `Minor`. All three resources, the resource message that describes what happened, and the file and line where the resource can be found, will be included in the event description.

Event severities can be configured via the `<change_type>_event_severity` class parameters.

You can specify the set of facts included in the event description via the `include_facts` parameter. It takes an array of strings that each represent a fact to retrieve from the available node's facts. Nested facts can be queried by dot notation such as `os.distro`, or `os.windows` to get the Windows version. Queries for nested facts must start at the top level fact name and any fact that is not present on a node, such as `os.windows` on a Linux box, is simply ignored without error.

Facts in the description by default are in Yaml format for readability, but this can be changed via the `facts_format` parameter to one of yaml, pretty_json (json with readability line breaks and indentation), or json.

To stop the module from sending any events to Servicenow, you can set the `disabled` parameter to `true`. Removing the module from classification on a Puppet server node will not stop the report processor from continuing to send events to Servicenow. To remove the integration permanently, uninstall the module from the environment and remove `servicenow` from the `reports` setting in [`puppet.conf`](https://puppet.com/docs/puppet/7/config_file_main.html).

### Incidents

To send incidents, classify your Puppet server nodes with the `servicenow_reporting_integration::incident_mangement` class. The minimum parameters you need to configure for the class to work are `instance` (the fqdn of the Servicenow instance), and then `user` and `password`, or you can bypass username/password authentication and use an oauth token using the `oauth_token` parameter. Lastly you will need to get the `sys_id` of the user you would like to use as the 'Caller' for each ticket. The `servicenow_reporting_integration::incident_management` class requires the `sys_id` for a user to be placed in the the `caller_id` parameter because that is a required incident field on ServiceNow’s end. Look below for steps to get the `sys_id` for a user from Servicenow.

To get the desired `sys_id` from Servicenow:
1. In the Application Navigator (left sidebar navigation menu) navigate to System Security > Users and Groups > Users
2. Use the search box to search for the user you want.
3. In the user listing click on a user.
4. In the user properties screen, click on the hamburger menu (three vertical bars) to the right of the left arrow in the upper left corner of the screen.
5. Click on 'Copy sys_id' to copy the `sys_id` directly to the clipboard, or click on Show XML to see the `sys_id` in a new window along with the rest of the user’s properties.

The `servicenow` report processor creates a ServiceNow incident based on a Puppet run report if at least one of the incident creation conditions are met.

By default the incident_creation_conditions are `['failures', 'corrective_changes']`. This means an incident will be created if there was a resource failure, a catalog compilation failure, or if a corrective change was made to a managed resource during a Puppet run. With these default parameters intentional changes such as adding a new resource to the catalog and Puppet bringing it into compliance for the first time, and pending changes, like running an agent in noop mode and the agent reporting changes would have occurred, will not result in an incident in Servicenow. If you would like to report on those types of changes `intentional_changes`, `pending_corrective_changes`, and `pending_intentional_changes` are also available as values for this parameter. As a shortcut you can also specify `always`, or to temporarily stop the module from creating any incidents at all you can set the `incident_creation_conditions` parameter to `never`.

The incident_mangement class also lets you specify additional (but optional) incident fields like the `category`, `subcategory` and `assigned_to` fields via the corresponding `category`, `subcategory`, and `assigned_to` parameters, respectively. See the `REFERENCE.md` for the full details.

Each incident will include the following information provided by Puppet:
* __Caller__: The user specified by the `sys_id` given to the `caller_id` parameter of the incident_management class. This can be any user in Servicenow and doesn’t need to be the same as the user that creates the incident via the username/password or oauth_token parameters.
* __Urgency__: Defaults to 3 - Low. Configurable
* __Impact__: Defaults to 3 - Low. Configurable
* __State__: Defaults to New. Configurable
* __Short Description__: Contains the following information
  * __Status__: changed, unchanged, failed, pending changes
  * __Node__: the fqdn of the node the agent ran on
  * __Report Time__: The timestamp of the report to help find the report in the console
* __Description__: See the description section from event management above. Incidents will get the same description

### Filtering

For both incidents and events, you can filter sending reports on the environment using an allow_list and block_list. You can set these lists in the module's parameters. The allow_list is defaulted to ['all'] and the block_list is defaulted to ['none']. If you have specific environment filters, you can add them to the list. You can also use wildcards (ie. ['*env_filter'], ['env_filter*'], ['*env_filter*']).

## Troubleshooting

To verify that everything worked, trigger a Puppet run on one of the nodes in the `PE Master` node group then log into the node. Once logged in, run `sudo tail -n 60 /var/log/puppetlabs/puppetserver/puppetserver.log | grep servicenow`. You should see some output. If not, then the class is probably not being classified properly. Either the class is not being assigned to the Puppet server nodes at all, or there may be catalog compilation errors on those nodes with the provided parameter values. Please use GitHub issues to file any bugs you find during your troubleshooting.

If you try to use the module and it doesn't work because of certificate validation errors, you can use the `skip_certificate_validation` parameter to disable certificate validation. The connection will still be SSL encrypted, but no certificate validation will be performed, which opens up a risk of 'Man in the middle' attacks. But, if you are using an internally hosted Servicenow instance that uses an SSL certificate that has not been imported for trust on the Puppet server machine, this may be necessary. Cloud hosted instances of Servicenow typically use a certificate signed by a well know public certificate authority, in which case, this parameter is not necessary.

If the module starts to throw errors indicating failures due to HTTP timeouts, you can try to use the `http_read_timeout` and `http_write_timeout` parameters to extend the amount of time available to complete a Servicenow API request.

If the module throws an error that contains the following text:

`failed to validate the PE console url 'https://<puppetserver URL>' via the '/auth/favicon.ico' endpoint: SSL_connect returned=1 errno=0 state=error: certificate verify failed`

This is an indication that the settings validation script cannot validate the availability and correct address of the pe console due to a certificate validation error. Please use the `$pe_console_cert_validation` parameter to change the method of certificate validation. By default the script assumes that the console is using the PE generate self signed cert (`selfsigned`). If you are using a certificate generated by an alternate means you will need to use either `truststore` if the ca cert that signed the console cert has been imported into the PE Servers OS trust store, or you can use `none` if you would like to skip certificate validate entirely. The connection will
still be SSL encrypted, but the certificate will not be validated.

## Development

### Unit tests
To run the unit tests:

```
bundle install
bundle exec rake spec_prep spec
```

### Acceptance tests
The acceptance tests use puppet-litmus in a multi-node fashion. The nodes consist of a 'master' node representing the PE master (and agent), and a 'ServiceNow' node representing the ServiceNow instance. All nodes are stored in a generated `inventory.yaml` file (relative to the project root) so that they can be used with Bolt.

To setup the test infrastructure, use `bundle exec rake acceptance:setup`. This will:

* **Provision the master VM**
* **Setup PE on the VM**
* **Setup the mock ServiceNow instance.** This is just a Docker container on the master VM that mimics the relevant ServiceNow endpoints. Its code is contained in `spec/support/acceptance/servicenow`.
* **Install the module on the master**

Each setup step is its own task; `acceptance:setup`'s implementation consists of calling these tasks. Also, all setup tasks are idempotent. That means its safe to run them (and hence `acceptance:setup`) multiple times.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. The format is based on [Keep a Changelog] and this project adheres to [Semantic Versioning].


## [Unreleased]

### Added

### Changed

### Fixed

### Removed

## [Release 0.2.2] - (2022-2-1)

### Added

- Add Environment filter ([PIE-523]) [#81]
- Add Report Filter for Events ([PIE-522]) [#78]

### Fixed

- Settings validation script failed with non-PE issued Console Certificate ([PIE-415]) [#71]
- Allow for the timeout parameters to accept either integer or float values ([PIE-998]) [#100]
- Prevent Trailing Slash in PE Console URL ([PIE-448]) [#72]

## [Release 0.1.1] - (2020-11-09)

### Fixed

- Fixed issue with event grouping on resource change events. ([PIE-409]) [#68]
- Added a check to `add_ignore_ok_events_rule` task to ensure variable is assigned before use. [#69]

## [Release 0.1.0] - (2020-10-27)

Initial Release

<!-- Reference links section -->

<!-- Ticket Links -->

[PIE-415]: https://tickets.puppetlabs.com/browse/PIE-415
[PIE-409]: https://tickets.puppetlabs.com/browse/PIE-409
[PIE-523]: https://tickets.puppetlabs.com/browse/PIE-523
[PIE-522]: https://tickets.puppetlabs.com/browse/PIE-522
[PIE-415]: https://tickets.puppetlabs.com/browse/PIE-415
[PIE-998]: https://tickets.puppetlabs.com/browse/PIE-998
[PIE-448]: https://tickets.puppetlabs.com/browse/PIE-448
[PIE-415]: https://tickets.puppetlabs.com/browse/PIE-415

<!-- Pull Request Links -->

[#68]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/68
[#69]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/69
[#71]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/71
[#81]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/81
[#78]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/78
[#100]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/100
[#72]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/pull/72

<!-- Version Comparison Links -->

[Unreleased]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/compare/v0.2.2...HEAD
[Release 0.2.2]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/compare/v0.1.1...v0.2.2
[Release 0.1.1]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/compare/v0.1.0...v0.1.1
[Release 0.1.0]: https://github.com/puppetlabs/puppetlabs-servicenow_reporting_integration/tree/v0.1.0

[Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: http://semver.org

spec/spec_helper_local.rb
# frozen_string_literal: true

RSpec.configure do |c|
  c.confdir = '.'
end


lib/puppet/util/servicenow.rb
require 'digest'
require 'puppet'
require 'puppet/util'
require 'fileutils'
require 'yaml'
require 'json'
require 'net/http'
require 'openssl'
require 'base64'

# hiera-eyaml requires. Note that newer versions of puppet-agent
# ship with the hiera-eyaml gem so these should work.
require 'hiera/backend/eyaml/options'
require 'hiera/backend/eyaml/parser/parser'
require 'hiera/backend/eyaml/subcommand'

# servicenow.rb
module Puppet::Util::Servicenow
  def sn_log_entry(msg)
    "servicenow report processor: #{msg}"
  end
  module_function :sn_log_entry

  def settings(settings_file = Puppet[:confdir] + '/servicenow_reporting.yaml')
    settings_hash = YAML.load_file(settings_file)

    # Since we also support hiera-eyaml encrypted passwords, we'll want to decrypt
    # the password before passing it into the request. In order to do that, we first
    # check if hiera-eyaml's configured on the node. If yes, then we run the password
    # through hiera-eyaml's parser. The parser will decrypt the password if it is
    # encrypted; otherwise, it will leave it as-is so that plain-text passwords are
    # unaffected.
    hiera_eyaml_config = nil
    begin
      # NOTE: If hiera-eyaml config doesn't exist, then load_config_file returns
      # the hash {:options => {}, :sources => []}
      hiera_eyaml_config = Hiera::Backend::Eyaml::Subcommand.load_config_file
    rescue StandardError => e
      raise "error reading the hiera-eyaml config: #{e}"
    end
    unless hiera_eyaml_config[:sources].empty?
      # hiera_eyaml config exists so run the password through the parser. Note that
      # we chomp the password to support syntax like:
      #
      #   password: >
      #       ENC[Y22exl+OvjDe+drmik2XEeD3VQtl1uZJXFFF2NnrMXDWx0csyqLB/2NOWefv
      #       NBTZfOlPvMlAesyr4bUY4I5XeVbVk38XKxeriH69EFAD4CahIZlC8lkE/uDh
      #       ...
      #
      # where the '>' will add a trailing newline to the encrypted password.
      #
      # Note that ServiceNow passwords can't contain a newline so chomping's still OK
      # for plain-text passwords.
      Hiera::Backend::Eyaml::Options.set(hiera_eyaml_config[:options])
      parser = Hiera::Backend::Eyaml::Parser::ParserFactory.hiera_backend_parser

      if (password = settings_hash['password'])
        password_tokens = parser.parse(password.chomp)
        password = password_tokens.map(&:to_plain_text).join
        settings_hash['password'] = password
      end

      if (oauth_token = settings_hash['oauth_token'])
        oauth_token_tokens = parser.parse(oauth_token.chomp)
        oauth_token = oauth_token_tokens.map(&:to_plain_text).join
        settings_hash['oauth_token'] = oauth_token
      end
    end

    settings_hash
  end
  module_function :settings

  def do_snow_request(uri, http_verb, body, user: nil, password: nil, oauth_token: nil, skip_cert_check: false, read_timeout: 60, write_timeout: 60)
    uri = URI.parse(uri)
    verify_mode = skip_cert_check ? OpenSSL::SSL::VERIFY_NONE : OpenSSL::SSL::VERIFY_PEER

    # We're going to set the connect timeout and the read timeout to the same
    # value, because in most user's minds these are functionally the same
    # timeout, and until we get some specific push for separating them out
    # we don't want to make the module more complex for users to understand.
    opts = {
      use_ssl:         uri.scheme == 'https',
      verify_mode:     verify_mode,
      read_timeout:    read_timeout,
      connect_timeout: read_timeout,
      ssl_timeout:     read_timeout,
      write_timeout:   write_timeout,
    }

    Net::HTTP.start(uri.host,
                    uri.port,
                    opts) do |http|
      header = { 'Content-Type' => 'application/json' }
      # Interpolate the HTTP verb and constantize to a class name.
      request_class_string = "Net::HTTP::#{http_verb.capitalize}"
      request_class = Object.const_get(request_class_string)
      # Add uri, fields and authentication to request
      request = request_class.new("#{uri.path}?#{uri.query}", header)
      request.body = body.to_json
      if oauth_token
        request['Authorization'] = "Bearer #{oauth_token}"
      else
        request.basic_auth(user, password)
      end
      # Make request to ServiceNow and return response
      http.request(request)
    end
  end
  module_function :do_snow_request

  def human_readable_event_summary(resource_statuses)
    summary = ''
    resource_statuses.values.select { |resource| resource.out_of_sync == true || resource.failed == true }.each do |resource|
      resource_summary = ''
      resource.events.each do |event|
        resource_summary << "#{['', resource.containment_path, "#{event.property}: #{event.message}"].flatten.join('/')}\n"
      end
      resource_summary << "  Resource Definition: #{resource.file}:#{resource.line}\n"
      summary << resource_summary
    end
    summary
  end
  module_function :human_readable_event_summary

  def additional_info_resource_events(resource_statuses)
    corrective_changes          = []
    intentional_changes         = []
    pending_corrective_changes  = []
    pending_intentional_changes = []
    failures                    = []
    resource_statuses.values.select { |resource| resource.out_of_sync == true || resource.failed == true }.each do |resource|
      event_summary = { 'resource'         => resource.resource,
                        'containing_class' => resource.resource_type,
                        'containment_path' => resource.containment_path,
                        'file'             => resource.file,
                        'line'             => resource.line }
      if resource.failed
        failures << event_summary
      elsif resource.events.count { |event| event.status == 'noop' } > 0
        (resource.corrective_change == true) ? (pending_corrective_changes << event_summary) : (pending_intentional_changes << event_summary)
      else
        (resource.corrective_change == true) ? (corrective_changes << event_summary) : (intentional_changes << event_summary)
      end
    end
    { 'corrective_changes'               => corrective_changes,
      'corrective_changes_hash'          => corrective_changes.empty? ? [] : Digest::SHA1.hexdigest(corrective_changes.to_json.chars.sort.join),
      'intentional_changes'              => intentional_changes,
      'intentional_changes_hash'         => intentional_changes.empty? ? [] : Digest::SHA1.hexdigest(intentional_changes.to_json.chars.sort.join),
      'pending_corrective_changes'       => pending_corrective_changes,
      'pending_corrective_changes_hash'  => pending_corrective_changes.empty? ? [] : Digest::SHA1.hexdigest(pending_corrective_changes.to_json.chars.sort.join),
      'pending_intentional_changes'      => pending_intentional_changes,
      'pending_intentional_changes_hash' => pending_intentional_changes.empty? ? [] : Digest::SHA1.hexdigest(pending_intentional_changes.to_json.chars.sort.join),
      'failures'                         => failures,
      'failures_hash'                    => failures.empty? ? [] : Digest::SHA1.hexdigest(failures.to_json.chars.sort.join) }
  end
  module_function :additional_info_resource_events

  # Returns a hash of event conditions
  def calculate_event_conditions(resource_statuses)
    event_conditions = {
      'failures'                    => false,
      'corrective_changes'          => false,
      'intentional_changes'         => false,
      'pending_corrective_changes'  => false,
      'pending_intentional_changes' => false,
    }

    resource_statuses.each_value do |resource|
      resource.events.each do |event|
        next if event.status == 'audit'
        # event.status == 'success' || 'noop'. Either way, we found a satisfying
        # change condition so determine its name
        if event.status == 'failure'
          change_condition = 'failures'
        else
          change_condition = event.corrective_change ? 'corrective_changes' : 'intentional_changes'
          change_condition = "pending_#{change_condition}" if event.status == 'noop'
        end
        event_conditions[change_condition] = true
      end
    end

    event_conditions
  end
  module_function :calculate_event_conditions

  def event_type(resource_statuses, status)
    case status
    when 'unchanged'
      'node_report_unchanged'
    when 'failed'
      'node_report_failed'
    when 'changed'
      event_conditions = calculate_event_conditions(resource_statuses)
      applicable_event_conditions = event_conditions.select { |_, condition| condition == true }.keys
      return 'node_report_corrective_changes' if applicable_event_conditions.include?('corrective_changes')
      'node_report_intentional_changes'
    end
  end
  module_function :event_type

  def event_severity_string(event_severity_string)
    severity_settings = {
      'Clear' => 0,
      'Critical' => 1,
      'Major' => 2,
      'Minor' => 3,
      'Warning' => 4,
      'OK' => 5,
    }
    severity_settings[event_severity_string]
  end

  def calculate_event_severity(resource_statuses, settings_hash, transaction_completed)
    # https://docs.servicenow.com/bundle/paris-it-operations-management/page/product/event-management-operator/concept/operator-events-alerts.html
    # 0 => Clear....(The alert no longer needs action.)
    # 1 => Critical.(The resource is either not functional or critical problems are imminent.)
    # 2 => Major....(Major functionality is severely impaired or performance has degraded.)
    # 3 => Minor....(Partial, non-critical loss of functionality or performance degradation occurred.)
    # 4 => Warning..(Attention is required, even though the resource is still functional.)
    # 5 => OK.......(No severity. An alert is created. The resource is still functional.)
    event_conditions = calculate_event_conditions(resource_statuses)
    # return no_changes_event_severity in the case that there are no changes
    return event_severity_string(settings_hash['failures_event_severity']) if catalog_compilation_failure?(resource_statuses, transaction_completed)
    return event_severity_string(settings_hash['no_changes_event_severity']) unless event_conditions.values.any?
    event_severity_string(event_conditions.select { |_, exists| exists == true }
                    .map { |condition, _| settings_hash[condition + '_event_severity'] }.sort.first)
  end
  module_function :calculate_event_severity

  # Returns an array of satisfied conditions
  # Note that the 'never' condition will be overridden by any other valid condition
  def calculate_satisfied_conditions(report_status, resource_statuses, incident_creation_conditions, transaction_completed)
    # Some incident creation conditions depend on the resource events. Thus, we go ahead
    # and evaluate these 'event' conditions _before_ evaluating the incident creation
    # conditions so we can iterate through the resource events only once. This results in
    # cleaner code for a negligible performance hit, where the performance hit occurs if the
    # incident creation conditions do _not_ contain an 'event' condition.
    #
    # Note that event_conditions is a hash of <condition> => <satisfied?>
    event_conditions = calculate_event_conditions(resource_statuses)
    satisfied_conditions = incident_creation_conditions.select do |condition|
      if condition == 'always'
        true
      elsif condition == 'never'
        false
      elsif condition == 'failures'
        report_status == 'failed' || catalog_compilation_failure?(resource_statuses, transaction_completed)
      elsif event_conditions.key?(condition)
        event_conditions[condition]
      else
        # We should never hit this code-path
        Puppet.warning(sn_log_entry("unknown incident creation conditon: #{condition}"))
        false
      end
    end
    Puppet.info(sn_log_entry("satisfied conditions: #{satisfied_conditions}"))
    satisfied_conditions
  end
  module_function :calculate_satisfied_conditions

  def report_description(settings_hash, resource_statuses, transaction_completed)
    resourse_status_summary = human_readable_event_summary(resource_statuses)
    labels                  = description_report_labels(resource_statuses, transaction_completed)
    # Ideally, we'd like to link to the specific report here. However, fine-grained PE console links are
    # unstable even for Y PE releases (e.g. the link is different for PE 2019.2 and PE 2019.8). Thus, the
    # best and most stable solution we can do (for now) is the description you see here.
    description =  labels.nil? ? '' : labels
    description << "\n\nEnvironment: #{environment}"
    description << "\n\nSee the PE console for the full report. You can access the PE console at #{settings_hash['pe_console_url']}."
    description << "\n\nResource Statuses:\n#{resourse_status_summary}" unless resourse_status_summary.empty?
    description << "\n\nLog Output:\n#{log_messages}" if catalog_compilation_failure?(resource_statuses, transaction_completed)
    description << "\n\n== Facts ==\n#{selected_facts(settings_hash)}"
    description
  end
  module_function :report_description

  def facts
    # This is a cheat to make it easier to do test mocks.
    Puppet::Node::Facts.indirection.find(host).values
  end

  def selected_facts(settings_hash, facts_query = nil, format = nil)
    include_facts = facts_query.nil? ? settings_hash['include_facts'] : facts_query
    output_format = format.nil? ? settings_hash['facts_format'].to_sym : format.to_sym

    selected_facts = {}

    if [include_facts].flatten.first == 'all'
      facts
    else
      include_facts.each do |name|
        value = facts.dig(*name.split('.'))
        selected_facts[name] = value unless value.nil?
      end
    end

    # json and object are primarily to support either internal uses like further
    # processing on selected facts, or giving to the facts to a machine
    # processor like an event engine.
    case output_format
    when :yaml
      selected_facts.to_yaml
    when :pretty_json
      JSON.pretty_generate(selected_facts)
    when :json
      selected_facts.to_json
    when :object
      selected_facts
    else
      selected_facts
    end
  end
  module_function :selected_facts

  def event_additional_information(settings_hash, resource_statuses, transaction_completed)
    additional_information = {}
    # If we wish to add other top level keys to the additional information field, add them here.
    # Include all facts since this field is not intended for humans.
    additional_information['environment'] = environment
    additional_information['report_labels'] = additional_info_report_labels(resource_statuses, transaction_completed)
    additional_information.merge!(selected_facts(settings_hash, nil, :object))
    additional_information.merge!(additional_info_resource_events(resource_statuses))
    JSON.pretty_generate(additional_information)
  end
  module_function :event_additional_information

  def format_report_timestamp(time, metrics)
    total_time = time + metrics['time']['total']
    short_date_time = total_time.strftime('%F %H:%M:%S %Z')
    short_date_time.gsub('UTC', 'Z')
  end
  module_function :format_report_timestamp

  def calculate_report_message_key_hash(report_status, resource_statuses)
    # The report message key hash consists of all the fields relevant to
    # determining a report's uniqueness. In our case, this is the report
    # status and the resource events.
    report_message_key_hash = {
      'status'          => report_status,
      'resource_events' => nil,
    }

    resource_events = []
    resource_statuses.each_value do |resource|
      resource.events.each do |event|
        # Some event fields are not relevant when it comes to determining
        # whether two reports are identical. We delete these unnecessary
        # event fields before adding the event.
        event_hash = event.to_data_hash
        event_hash.delete('historical_value')
        event_hash.delete('message')
        event_hash.delete('time')
        event_hash.delete('redacted')

        resource_events << event_hash
      end
    end
    report_message_key_hash['resource_events'] = resource_events

    report_message_key_hash
  end
  module_function :calculate_report_message_key_hash

  def report_labels(resource_statuses, transaction_completed)
    event_conditions = calculate_event_conditions(resource_statuses).select { |_, present| present == true }
    labels = event_conditions.keys
    labels << 'catalog_failure' if catalog_compilation_failure?(resource_statuses, transaction_completed)
    labels
  end
  module_function :report_labels

  def description_report_labels(resource_statuses, transaction_completed)
    labels = report_labels(resource_statuses, transaction_completed)
    "Report Labels:\n\t#{labels.join("\n\t")}" unless labels.empty?
  end
  module_function :description_report_labels

  def additional_info_report_labels(resource_statuses, transaction_completed)
    labels = report_labels(resource_statuses, transaction_completed)
    labels.join(', ')
  end
  module_function :additional_info_report_labels

  def log_messages
    logs.map { |entry| entry.message }.join("\n")
  end
  module_function :log_messages

  def catalog_compilation_failure?(resource_statuses, transaction_completed)
    resource_statuses.empty? && !transaction_completed
  end
  module_function :catalog_compilation_failure?

  def environment_matched?(environment, allow, block)
    return true if allow == 'none' || block == 'all'
    return false if block == 'none' || allow == 'all'
    return true if File.fnmatch(block, environment)
    return false if File.fnmatch(allow, environment)
  rescue e
    raise e.message, 'Error with environment filter. Please look over allow_list and block_list configurations'
  end
  module_function :environment_matched?

  def env_filter_not_allowed?(environment, allow_list, block_list)
    blocked = false
    if allow_list.length >= block_list.length
      allow_list.zip(block_list) do |allow, block|
        blocked = environment_matched?(environment, allow, block)
      end
    elsif block_list.length > allow_list.length
      block_list.zip(allow_list) do |block, allow|
        blocked = environment_matched?(environment, block, allow)
      end
    end
    Puppet.info(sn_log_entry("Environment filter not allowed: #{blocked}"))
    blocked
  end
  module_function :env_filter_not_allowed?
end

# takes the instance string from the settings and prepends 'https://' if not already present.
def instance_with_protocol(instance)
  if instance[0..7] == 'https://'
    instance
  else
    'https://' << instance
  end
end
